### PR TITLE
Initialize rules from XML files

### PR DESCRIPTION
**Pull request by cedric-anne**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Based on #13234.

Rules are now initialized using XML files. It does not share logic with rules import process as the import process is too complex to be used during initialization. Indeed, here we do not have to check entities/dropdown existence (rules are all defined on root entity and do not use dropdown in criteria/actions).

I made folowing changes on `RuleImportAsset` default rules:
- ranking now starts at `1` instead of `0`,
- rules are now recursive.

XML files are produced using the export feature. I just removed values in `date_creation` nodes.

### CODE REVIEW

**Review by stonebuzz — APPROVED** [review]

**Review by trasher — APPROVED** [review]

File: src/Rule.php [hunk 1]
```diff
@@ -3553,4 +3553,163 @@ public function prepareInputForClone($input)
...
+            }
+            if ((string)$rulexml->entities_id !== 'Root entity') {
+                trigger_error(sprintf('Unexpected entity value for rule `%s`.', (string)$rulexml->uuid), E_USER_WARNING);
+                $has_errors = true;
+                continue;
+            }
+
+            $rule = new static();
+
+            if ($check === true && $rule->getFromDBByCrit(['uuid' => (string)$rulexml->uuid])) {
```
Comment (orthagh) [inline src/Rule.php:3619]: as we re-init the rules, could it be better to remove the old rule and re-insert it from the file ?

File: src/RuleCollection.php [hunk 2]
```diff
@@ -669,9 +669,10 @@ public function showListRules($target, $options = [])
             $url = $CFG_GLPI["root_doc"];
         }
 
-       //if rules provides an initRules method, then we're able to reset them
-        if (method_exists($this->getRuleClass(), 'initRules')) {
-            echo "<a class='btn btn-primary' id='reset_rules' href='" . $rule->getSearchURL() . "?reinit=true&subtype=" . $this->getRuleClassName() . "' " .
+        // if rules provides has default rules, then we're able to reset them
+        $ruleclass = $this->getRuleClass();
+        if ($ruleclass instanceof Rule && $ruleclass->hasDefaultRules()) {
```
Comment (orthagh) [inline src/RuleCollection.php:674]: instanceof test required ? I though getRuleClass method will always return a Rule class, no ?

Comment (cedric-anne) [reply src/RuleCollection.php:674]: Result could be `null` too.

File: src/Rule.php (hunk 1 above)
Comment (cedric-anne) [reply src/Rule.php:3619]: `$check` param is set to `false` in our case, and does not seems to be used. I keep it to not introduce a BC break, and just reuse the same logic as previous code, but with simplier criteria.
image

**Review by orthagh — APPROVED** [review]

**Pull request merged**